src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,401 @@
+ */
+
+package javax.crypto.spec;
+
+import javax.crypto.SecretKey;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Parameters for the combined Extract-Only, Expand-Only, or Extract-then-Expand operations of the
+ * HMAC-based Key Derivation Function (HKDF). The HKDF function is defined in <a
+ * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+ *
+ * @since 23
+ */
+public interface HKDFParameterSpec extends KDFParameterSpec {
+
+    /**
+     * This builder helps with the mutation required by the {@code Extract} scenario.
+     */
+    final class Builder {
+
+        Extract extract = null;
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+        SecretKey prk = null;
+
+        Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createExtract() {
+            extract = new Extract();
+            return this;
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an {@code Extract}
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            if (this.ikms.isEmpty() && this.salts.isEmpty()) {
+                throw new IllegalStateException(
+                    "this `Builder` must have either at least one IKM value, at least one salt "
+                    + "value, or values for both before calling `extractOnly`");
+            } else {
+                this.extract = new Extract(List.copyOf(ikms), List.copyOf(salts));
+                return this.extract;
+            }
+        }
+
+        /**
+         * Akin to a {@code Builder.build()} method for an {@code ExtractExpand}
+         *
+         * @param info
+         *     the info
+         * @param length
+         *     the length
+         *
+         * @return an {@code ExtractExpand}
+         */
+        public ExtractExpand andExpand(byte[] info, int length) {
+            if (extract == null) {
+                throw new IllegalStateException(
+                    "`thenExpand` can only be called on a `Builder` when `ofTypeExtract` has "
+                    + "been called");
+            }
+            return extractExpand(new Extract(List.copyOf(ikms), List.copyOf(salts)), info, length);
+        }
+
+        /**
+         * {@code addIKM} may be called when the ikm value is to be assembled piece-meal or if part
+         * of the IKM is to be supplied by a hardware crypto device. This method appends to the
+         * existing list of values or creates a new list if there are none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM but the actual value
+         * of the IKM is not yet available.
+         *
+         * @param ikm
+         *     the ikm value (null values will not be added)
```

Comment:
Are you sure about allowing `null`? Any valid use case? Same question for the other `add` methods.

Reply:
`null` should not be silently ignored. The method should throw `NullPointerException` (and that should be added to the javdoc in an `@throws` clause).